### PR TITLE
Reenable popcnt for MSVC, improve perf by leveraging ILP

### LINKED ISSUE #36: MSVC hardware-assisted popcount implementation
In #25 support was added for hardware-assisted popcount.
In #33 it was noted that the implementation for MSVC failed to check for CPU support, and the MSVC documentation states the behavior is undefined in this case.
In #35 the MSVC implementation was disabled to resolve #33, which leaves the library performance in this respect the same as it was in 1.68.0.

This issue is a placeholder to come back around and clean up the implementation so it can leverage hardware assist safely with MSVC.  There are some thoughts in the comments for #33 regarding alternate implementations that are useful when making a decision here.

### PR DESCRIPTION
**Pull request by Lastique**
On MSVC and compatible compilers popcnt is now enabled if it is known at
compile time that the target CPU supports popcnt instruction. Unfortunately,
MSVC does not have a predefined macro specifically for popcnt, so we have
to test for AVX instead. `__POPCNT__` is still tested in case clang or user
defines it.

For MSVC, the 64-bit popcount is now also implemented on 32-bit targets by
issuing two 32-bit popcnt instructions. For gcc and compatible compilers,
16-bit `popcount` implementation is provided, which is better than the generic
`popcount` implementation, but still slower than byte-wise `do_count`.

The `do_count` algorithm implementations have been improved by leveraging
instruction level parallelism better, which gives 0 to 27% improvement and
no regressions.

Also made code formatting more consistent and reduced code duplication between
`do_count` implementations.

This fixes #36

### CODE REVIEW

**Comment by jeking3** [comment]
I know about this, it's behind refreshing the CI to clear out the build failure(s).

**Comment by jeking3** [comment]
@Lastique please rebase on develop to clear out the build issues.  Thanks.

**Comment by jeking3** [comment]
At this point I think the failures are legit and not a product of CI.

**Comment by Lastique** [comment]
What failures?

**Comment by jeking3** [comment]
Click through to the Appveyor build.

**Comment by Lastique** [comment]
I'll take a look later, thanks.

**Comment by Lastique** [comment]
The tests have passed.

File: include/boost/dynamic_bitset/detail/dynamic_bitset.hpp [hunk 1]
```diff
@@ -85,126 +86,129 @@ namespace boost {
...
+    //
+    template <typename Iterator, typename ValueType>
+    inline std::size_t do_count(Iterator first, std::size_t length, ValueType,
+                                value_to_type<access_by_blocks>*)
+    {
+        std::size_t num1 = 0u, num2 = 0u;
+        while (length >= 2u) {
+            num1 += popcount<ValueType>(*first);
+            ++first;
+            num2 += popcount<ValueType>(*first);
```
Comment (jeking3) [inline include/boost/dynamic_bitset/detail/dynamic_bitset.hpp]: It's interesting this improves performance.  Simply by assigning the result two two variables (i.e. registers) that allows them to execute independently even though they are both tied to the advancement of the iterator?  I would have expected to need some sort of hint or intrinsic for that.  Is it the compiler or the CPU doing this?

Comment (jeking3) [reply include/boost/dynamic_bitset/detail/dynamic_bitset.hpp]: I guess on Intel architecture it's CPU based, after some reading.

Comment (Lastique) [reply include/boost/dynamic_bitset/detail/dynamic_bitset.hpp]: The iterator is most likely advanced only once per loop iteration. One of the two iterator dereference operations translates to a memory access with a static offset.

The important optimization here is breaking one dependency chain on the accumulator into two, which allows the CPU to execute popcnt and accumulate instructions in parallel for the two dependency chains. Instruction-level parallelism is a common feature of most modern CPUs that support out-of-order execution. Theoretically, a compiler could generate an equivalent code, but I haven't seen a compiler do this.

Comment (jeking3) [reply include/boost/dynamic_bitset/detail/dynamic_bitset.hpp]: Potentially any loop unroll without side-effects would be eligible for this sort of thing?

Comment (Lastique) [reply include/boost/dynamic_bitset/detail/dynamic_bitset.hpp]: Yes, most of the time. You may not get any speedup in some cases, e.g. if your algorithm is memory-bound or the loop body becomes too large for the out-of-order window in the CPU. The latter case is somewhat mitigated by instruction reordering done by compiler or by hand.

Comment (jeking3) [reply include/boost/dynamic_bitset/detail/dynamic_bitset.hpp]: Sounds like an interesting compiler option if you ask me. :)  Thanks for the explanation.

File: include/boost/dynamic_bitset/detail/dynamic_bitset.hpp [hunk 2]
```diff
@@ -85,126 +86,129 @@ namespace boost {
...
+    // Some platforms have fast popcount operation, that allow us to implement
+    // counting bits much more efficiently
+    //
+    template <typename ValueType>
+    BOOST_FORCEINLINE std::size_t popcount(ValueType value) BOOST_NOEXCEPT
+    {
+        std::size_t num = 0u;
+        while (value) {
+            num += count_table<>::table[value & ((1u<<table_width) - 1)];
+            value >>= table_width;
```
Comment (jeking3) [inline include/boost/dynamic_bitset/detail/dynamic_bitset.hpp]: Coverity flagged this issue:

CID 338011 (#1 of 1): Bad bit shift operation (BAD_SHIFT)
2. large_shift: In expression value >>= 8U, right shifting value by more than 7 bits always yields zero. The shift amount is 8.

I believe this is intentional in this case, but I want to double-check with you.

Comment (Lastique) [reply include/boost/dynamic_bitset/detail/dynamic_bitset.hpp]: 1. I believe the original code was the same. The change was in the indentation.
2. The intention of the code is to eventually shift off all set bits, leaving `value` zero. If `value` is less than `int`, it is promoted to `int` before the shift operation, so in no case the shift by 8 bits exceeds the left hand operand capacity. I believe, the error is wrong.

Comment (jeking3) [reply include/boost/dynamic_bitset/detail/dynamic_bitset.hpp]: I'll mark it as such, thanks.  There's another interesting one in the issues which looks like a real bug.

**Pull request merged**